Remove optional parameters on updater and view functions

warrenseymour opened the pull request:
By marking parameters as optional, TypeScript requires verbose guards when running in strict mode. I do not think these parameters should be marked as optional, as Casium will always supply them internally. However, it is up to the implementer to actually use them.

By marking them as non-optional, type checking and completion when writing updater and view functions is a much more pleasant experience.

In order to achieve this, the `UpdaterDef` and `ContainerViewDef` type aliases have been introduced, which mirror their existing counterparts but do not contain optional parameters. These types are presented to consumerrs of `container` when writing Updater and View functions.

This PR also adds support for function-returning Updaters to `UpdateResult` - since an Updater can now return a function (see e9712ee9722d3988f1ad2511c35e1c5bb4377aab), it follows that the definition of `UpdateResult` should also include this as a permissible type.

nateabele commented on src/core.ts:
```diff
@@ -191,8 +191,8 @@ export const isolate = <M>(ctr: Container<M>, opts: any = {}): IsolatedContainer
-  return function (model: M, msg?: GenericObject, relay?: GenericObject): UpdateResult<M> {
-    const merge = ([{}, cmds], [newModel, newCmds]) => [newModel, flatten(cmds.concat(newCmds))];
+  return function(model: M, msg: GenericObject, relay: GenericObject): UpdateResult<M> {
+    const merge = ([{ }, cmds], [newModel, newCmds]) => [newModel, flatten(cmds.concat(newCmds))];
```
The spacing here is inconsistent with the coding standard. Should be `return function (` and `merge = ([{}`.

nateabele commented:
Looks like the tests need to be updated to account for the stricter type constraints.

warrenseymour commented:
@nateabele Rebased with whitespace changes removed and tests updated

nateabele commented on src/core_test.tsx:
```diff
@@ -95,12 +95,13 @@ describe('app', () => {
       const ctr = container({ name: 'FooContainer', update, view });
 
-      expect(() => shallow(ctr()).html()).to.throw(
+      expect(() => shallow(ctr(state)).html()).to.throw(
```
Sorry, it didn't even occur to me before, but this actually has a pretty big impact on how tests are written, so `isolate()` will need to be patched to handle passing required parameters internally.

warrenseymour commented:
Ah yes, good point.

On Fri, Feb 23, 2018 at 4:06 PM, Nate Abele <notifications@github.com>
wrote:

> *@nateabele* commented on this pull request.
> ------------------------------
>
> In src/core_test.tsx
> <https://github.com/ai-labs-team/casium/pull/36#discussion_r170291752>:
>
> >
>      it('throws on unhandled messages', () => {
>        const view = ({ emit }) => (<button onClick={emit(Msg2)}>Foo</button>);
>        const ctr = container({ name: 'FooContainer', update, view });
>
> -      expect(() => shallow(ctr()).html()).to.throw(
> +      expect(() => shallow(ctr(state)).html()).to.throw(
>
> Sorry, it didn't even occur to me before, but this actually has a pretty
> big impact on how tests are written, so isolate() will need to be patched
> to handle passing required parameters internally.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ai-labs-team/casium/pull/36#pullrequestreview-98961697>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAogpHxMstRpmI1kGVsf-QtC9Luu7DbCks5tXuHvgaJpZM4SPoXt>
> .
>

warrenseymour commented:
@nateabele I've updated this PR and changed its base to the `next` branch. As you will see, by adding some type aliases I've been able to introduce the desired effect without needing to change the tests.
